support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go

```diff
@@ -0,0 +1,177 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/v2/svid/x509svid"
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
+	"github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	certsFileMode      = os.FileMode(0644)
+	keyFileMode        = os.FileMode(0600)
+	certsFileName      = "tls.crt"
+	keyFileName        = "tls.key"
+	trustBundleTimeout = 3 * time.Minute
+)
+
+type SVIDWatcherConfig struct {
+	AgentSocketPath string
+	Ctx             context.Context
+	Log             logrus.FieldLogger
+	SpiffeID        *types.SPIFFEID
+	WebhookCertDir  string
+}
+
+type SVIDWatcher struct {
+	c SVIDWatcherConfig
+}
+
+func NewSVIDWatcher(config SVIDWatcherConfig) *SVIDWatcher {
+	return &SVIDWatcher{
+		c: config,
+	}
+}
+
+// startWatcher sets up the process to download and rotate the webhook certificates
+func (s *SVIDWatcher) Start() error {
+	err := os.MkdirAll(s.c.WebhookCertDir, 0700)
+	if err != nil {
+		return err
+	}
+
+	client, err := workloadapi.New(s.c.Ctx, workloadapi.WithAddr("unix://"+s.c.AgentSocketPath))
+	if err != nil {
+		return fmt.Errorf("unable to create workload API client: %w", err)
+	}
+
+	w := &x509Watcher{
+		s:      s,
+		svidCh: make(chan *x509svid.SVID),
+	}
+
+	go func() {
+		defer client.Close()
+		err := client.WatchX509Context(s.c.Ctx, w)
+		if err != nil && status.Code(err) != codes.Canceled {
+			s.c.Log.Fatalf("Error watching X.509 context: %w", err)
+		}
+	}()
+
+	// Wait for the initial trust bundle to arrive. We can't start the webhook without the certs in place.
+	select {
```

Comment:
Also need to select on s.c.Ctx here in case the outer context is canceled before the three minutes are up.